README rewrite for DU functions (#222)
streamr-dev/streamr-client-javascript: README.md

## Data Unions

This library provides functions for working with Data Unions.

> To deploy a new DataUnion:

Reviewer: Is this mixed up with 
**_To get an existing (previously deployed) `DataUnion` instance:_**

Author: good catch

```js

> const dataUnion = client.getDataUnion(address)

Reviewer 2: This example codes and their descriptions don't match

Author: fixed

## Data Unions

This library provides functions for working with Data Unions.

To deploy a new DataUnion:
```js
const dataUnion = await client.deployDataUnion()
```

To get an existing (previously deployed) `DataUnion` instance:
```js

> const dataUnion = client.getDataUnion(address)

Reviewer: ```suggestion
const dataUnion = client.getDataUnion(dataUnionAddress)
```

| createSecret(\[name])             | string              | Create a secret for a Data Union                               |
| setAdminFee(newFeeFraction)       | Transaction receipt | `newFeeFraction` is a `Number` between 0.0 and 1.0 (inclusive) |
| addMembers(memberAddressList)     | Transaction receipt | Add members                                                    |
| removeMembers(memberAddressList)  | Transaction receipt | Remove members from Data Union                                 |
| withdrawAllToMember(memberAddress\[, [options](#withdraw-options)\])                              | Transaction receipt | Send all withdrawable earnings to the member's address |
| withdrawAllToSigned(memberAddress, recipientAddress, signature\[, [options](#withdraw-options)\]) | Transaction receipt | Send all withdrawable earnings to the address signed off by the member (see [example below](#member-functions)) |

Here's how to deploy a Data Union contract and set the admin fee to 30%:

```js
import { StreamrClient } from 'streamr-client'

const client = new StreamrClient({
    auth: { privateKey },
})

const dataUnion = await client.deployDataUnion()

> const receipt = await dataUnion.setAdminFee(0.3, { dataUnion })

Reviewer: ```suggestion
const receipt = await dataUnion.setAdminFee(0.3)
```

| join(\[secret])                                                   | JoinRequest         | Join the Data Union (if a valid secret is given, the promise waits until the automatic join request has been processed)  |
| isMember(memberAddress)                                           | boolean             |                                                                             |
| withdrawAll(\[[options](#withdraw-options)\])                     | Transaction receipt | Withdraw funds from Data Union                                              |
| withdrawAllTo(recipientAddress\[, [options](#withdraw-options)\]) | Transaction receipt | Donate/move your earnings to recipientAddress instead of your memberAddress |
| signWithdrawAllTo(recipientAddress)                               | Signature (string)  | Signature that can be used to withdraw all available tokens to given recipientAddress        |
| signWithdrawAmountTo(recipientAddress, amountTokenWei)            | Signature (string)  | Signature that can be used to withdraw a specific amount of tokens to given recipientAddress |

> Here's an example how to sign off on a withdraw to (any) recipientAddress (NOTE: this requires no gas!)

Reviewer: ```suggestion
Here's an example on how to sign off on a withdraw to (any) recipientAddress (NOTE: this requires no gas!)
```

```js
import { StreamrClient } from 'streamr-client'

const client = new StreamrClient({
    auth: { privateKey },
})

const dataUnion = client.getDataUnion(dataUnionAddress)
const signature = await dataUnion.signWithdrawAllTo(recipientAddress)
```

Later, anyone (e.g. Data Union admin) can send that withdraw transaction to the blockchain (and pay for the gas)

```js
import { StreamrClient } from 'streamr-client'

const client = new StreamrClient({
    auth: { privateKey },
})

> const dataUnion = client.getDataUnion(dataUnionAddress)

Reviewer: ```suggestion
const dataUnion = client.getDataUnion(address)
```
for consistency,  dataUnionAddress -> address

> Here's an example how to deploy a data union contract and set the admin fee:

Reviewer: ```suggestion
Here's how to deploy a Data Union contract and set the admin fee to 30%:
```

```js
import { StreamrClient } from 'streamr-client'

const client = new StreamrClient({
    auth: { privateKey },
})

const dataUnion = await client.deployDataUnion()

> const receipt = await dataUnion.setAdminFee(0.3, { dataUnion })

Reviewer 2: The second parameter `(({ dataUnion })` has been removed

Author: it has in fact been removed  ;)  https://github.com/streamr-dev/streamr-client-javascript/pull/224

Author: or about to be

Reviewer 2: In #224 you remove option for another method (`addMembers`, `removeMembers`), not `setAdminFee`.

Reviewer: Fixed here

Here's an example how to get a member's withdrawable token balance (in "wei", where 1 DATA = 10^18 wei)

```js
import { StreamrClient } from 'streamr-client'

> const dataUnion = new StreamrClient().getDataUnion(dataUnionAddress)

Reviewer: ```suggestion
const dataUnion = new StreamrClient().getDataUnion(address)
```

const withdrawableWei = await dataUnion.getWithdrawableEarnings(memberAddress)
```

### Withdraw options

> The functions `withdrawAll`, `withdrawAllTo`, `withdrawAllToMember`, `withdrawAllToSigned` all take an extra "options" argument. It's an object can contain following members:

Reviewer: ```suggestion
The functions `withdrawAll`, `withdrawAllTo`, `withdrawAllToMember`, `withdrawAllToSigned` all take an extra "options" argument. It's an object that can contain the following parameters:
```

Reviewer: Avoiding 'members' to avoid confusion with DU members

Publishing to npm is automated via Github Actions. Follow the steps below to publish `latest` or `beta`.

> These docs can be rebuilt locally via:

Reviewer: ```suggestion
For more technical documentation on the Data Unions API, see the [JS Client API Docs](https://streamr-dev.github.io/streamr-client-javascript/). These can also be rebuilt locally via:
```
